[ao] Created Skeleton for ModelReportVisualizer class

vspenubarthi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #81673
* #81647
* #81589
* __->__ #81523

Summary: This introduces the skeleton for the ModelReportVisualizer
class. This class helps visualize the information generated by the
ModelReport class `generate_report()` output. This class aims to provide
visualizations in a table, plot (line graph) and histogram view.

This also introduces an empty test class for testing visualizations. As
implementations start occuring for this class, tests will also be
approrpriately added.

This includes the high level descriptions for each of the methods as
well. Expected use cases will be added to the class description in a
future commit as that gets finalized.

Test Plan: python test/test_quantization.py TestFxModelReportVisualizer

Reviewers:

Subscribers:

Tasks:

Tags:

vspenubarthi committed:
[ao] Created Skeleton for ModelReportVisualization class

Summary: This introduces the skeleton for the ModelReportVisualization
class. This class helps visualize the information generated by the
ModelReport class `generate_report()` output. This class aims to provide
visualizations in a table, plot (line graph) and histogram view.

This also introduces an empty test class for testing visualizations. As
implementations start occuring for this class, tests will also be
approrpriately added.

This includes the high level descriptions for each of the methods as
well. Expected use cases will be added to the class description in a
future commit as that gets finalized.

Test Plan: python test/test_quantization.py TestFxModelReportVisualization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi commented:
DO NOT REVIEW JUST YET, CHANGES COMING

vspenubarthi committed:
Update on "[ao] Created Skeleton for ModelReportVisualization class"


Summary: This introduces the skeleton for the ModelReportVisualization
class. This class helps visualize the information generated by the
ModelReport class `generate_report()` output. This class aims to provide
visualizations in a table, plot (line graph) and histogram view.

This also introduces an empty test class for testing visualizations. As
implementations start occuring for this class, tests will also be
approrpriately added.

This includes the high level descriptions for each of the methods as
well. Expected use cases will be added to the class description in a
future commit as that gets finalized.

Test Plan: python test/test_quantization.py TestFxModelReportVisualization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Created Skeleton for ModelReportVisualization class"


Summary: This introduces the skeleton for the ModelReportVisualization
class. This class helps visualize the information generated by the
ModelReport class `generate_report()` output. This class aims to provide
visualizations in a table, plot (line graph) and histogram view.

This also introduces an empty test class for testing visualizations. As
implementations start occuring for this class, tests will also be
approrpriately added.

This includes the high level descriptions for each of the methods as
well. Expected use cases will be added to the class description in a
future commit as that gets finalized.

Test Plan: python test/test_quantization.py TestFxModelReportVisualization

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Created Skeleton for ModelReportVisualizer class"


Summary: This introduces the skeleton for the ModelReportVisualizer
class. This class helps visualize the information generated by the
ModelReport class `generate_report()` output. This class aims to provide
visualizations in a table, plot (line graph) and histogram view.

This also introduces an empty test class for testing visualizations. As
implementations start occuring for this class, tests will also be
approrpriately added.

This includes the high level descriptions for each of the methods as
well. Expected use cases will be added to the class description in a
future commit as that gets finalized.

Test Plan: python test/test_quantization.py TestFxModelReportVisualizer

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Created Skeleton for ModelReportVisualizer class"


Summary: This introduces the skeleton for the ModelReportVisualizer
class. This class helps visualize the information generated by the
ModelReport class `generate_report()` output. This class aims to provide
visualizations in a table, plot (line graph) and histogram view.

This also introduces an empty test class for testing visualizations. As
implementations start occuring for this class, tests will also be
approrpriately added.

This includes the high level descriptions for each of the methods as
well. Expected use cases will be added to the class description in a
future commit as that gets finalized.

Test Plan: python test/test_quantization.py TestFxModelReportVisualizer

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Created Skeleton for ModelReportVisualizer class"


Summary: This introduces the skeleton for the ModelReportVisualizer
class. This class helps visualize the information generated by the
ModelReport class `generate_report()` output. This class aims to provide
visualizations in a table, plot (line graph) and histogram view.

This also introduces an empty test class for testing visualizations. As
implementations start occuring for this class, tests will also be
approrpriately added.

This includes the high level descriptions for each of the methods as
well. Expected use cases will be added to the class description in a
future commit as that gets finalized.

Test Plan: python test/test_quantization.py TestFxModelReportVisualizer

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi committed:
Update on "[ao] Created Skeleton for ModelReportVisualizer class"


Summary: This introduces the skeleton for the ModelReportVisualizer
class. This class helps visualize the information generated by the
ModelReport class `generate_report()` output. This class aims to provide
visualizations in a table, plot (line graph) and histogram view.

This also introduces an empty test class for testing visualizations. As
implementations start occuring for this class, tests will also be
approrpriately added.

This includes the high level descriptions for each of the methods as
well. Expected use cases will be added to the class description in a
future commit as that gets finalized.

Test Plan: python test/test_quantization.py TestFxModelReportVisualizer

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -0,0 +1,160 @@
+        """
+        pass
+
+    def get_all_unique_feature_names(self, plottable: bool) -> Set[str]:
```
same for this one

vspenubarthi replied:
for now, keeping this here because it was envisioned that this would be a tool that would help the users use the filters. Another thing is that currently, **the format for the `generated reports` is not the same as the format for the reports stored in the ModelReportVisualizer. This commit https://github.com/pytorch/pytorch/pull/81589 shows in the implementation that there is a transformation done to go from the reports in ModelReport (which are detector by detector) to the ones here, which are module by module.

vspenubarthi replied:
will revisit as it gets fleshed out more to see if it makes sense to move it to the ModelReport class, but makes sense being here to me.

jerryzh168 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -0,0 +1,160 @@
+        """
+        pass
+
+    def generate_table_view(self, feature: str = "", module_fqn_prefix_filter: str = "") -> Tuple[List[List[Any]], str]:
```
is this really a visualization? or just reorganizing the reported results based on some args?

vspenubarthi replied:
this does end up being a visualization as well. The two returned things are:
a list of lists format of the filtered information.
A string that they can just print to get the visualization. Implementation is here: https://github.com/pytorch/pytorch/pull/81673

jerryzh168 reviewed (commented):
had a few questions inline

andrewor14 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -0,0 +1,160 @@
+    The ModelReportVisualizer class aims to provide users a way to visualize some of the statistics
+    that were generated by the ModelReport API. However, at a higher level, the class aims to provide
+    some level of visualization of statistics to PyTorch in order to make it easier to parse data and
+    diagnose any potential issues with data or a specific model. With respects to the visualizations,
```
respects -> respect

andrewor14 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -0,0 +1,160 @@
+        """
+        pass
+
+    def generate_table_view(self, feature: str = "", module_fqn_prefix_filter: str = "") -> Tuple[List[List[Any]], str]:
```
We should make the return value consistent with that of `generate_detector_report`, where the string is the first item of the tuple, not the second

andrewor14 commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -0,0 +1,160 @@
+        """
+        pass
+
+    def generate_plot_view(self, feature: str, module_fqn_prefix_filter: str = "") -> List[List[Any]]:
```
Why does `generate_table_view` return an info string but not `generate_plot_view` and `generate_histogram_view`? I feel the return signatures of all these should be the same

vspenubarthi replied:
the thing that all three return consistently is the list of lists representing the filtered data for them to go through. The reason `generate_table_view` returns a string is because we want to present the table in a string format so they can just print it see [this line](https://github.com/pytorch/pytorch/pull/81523/files/92227b0ffa3eb62d878679af8f60e67bd282689f#diff-5d4780a142975c963b00de031975b2b538df9ff8a069d3b43d7c4099165fb5dfR113) for expected usage.

Meanwhile, **I agree about the inconsistency with the histogram and plot, so I update those to be in the same tuple** format, but they return functions that can be used plot the graphs instead of strings to print

vspenubarthi committed:
Update on "[ao] Created Skeleton for ModelReportVisualizer class"


Summary: This introduces the skeleton for the ModelReportVisualizer
class. This class helps visualize the information generated by the
ModelReport class `generate_report()` output. This class aims to provide
visualizations in a table, plot (line graph) and histogram view.

This also introduces an empty test class for testing visualizations. As
implementations start occuring for this class, tests will also be
approrpriately added.

This includes the high level descriptions for each of the methods as
well. Expected use cases will be added to the class description in a
future commit as that gets finalized.

Test Plan: python test/test_quantization.py TestFxModelReportVisualizer

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi commented:
Ready for review now btw

vspenubarthi commented on torch/ao/quantization/fx/_model_report/model_report_visualizer.py:
```diff
@@ -0,0 +1,160 @@
+        """
+        pass
+
+    def get_all_unique_module_fqns(self) -> Set[str]:
```
for now, keeping this here because it was envisioned that this would be a tool that would help the users use the filters. Another thing is that currently, **the format for the `generated reports` is not the same as the format for the reports stored in the ModelReportVisualizer. This commit https://github.com/pytorch/pytorch/pull/81589 shows in the implementation that there is a transformation done to go from the reports in ModelReport (which are detector by detector) to the ones here, which are module by module.

vspenubarthi replied:
will revisit as it gets fleshed out more to see if it makes sense to move it to the ModelReport class, but makes sense being here to me.

vspenubarthi committed:
Update on "[ao] Created Skeleton for ModelReportVisualizer class"


Summary: This introduces the skeleton for the ModelReportVisualizer
class. This class helps visualize the information generated by the
ModelReport class `generate_report()` output. This class aims to provide
visualizations in a table, plot (line graph) and histogram view.

This also introduces an empty test class for testing visualizations. As
implementations start occuring for this class, tests will also be
approrpriately added.

This includes the high level descriptions for each of the methods as
well. Expected use cases will be added to the class description in a
future commit as that gets finalized.

Test Plan: python test/test_quantization.py TestFxModelReportVisualizer

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi commented:
@pytorchbot merge -g

vspenubarthi committed:
Update on "[ao] Created Skeleton for ModelReportVisualizer class"


Summary: This introduces the skeleton for the ModelReportVisualizer
class. This class helps visualize the information generated by the
ModelReport class `generate_report()` output. This class aims to provide
visualizations in a table, plot (line graph) and histogram view.

This also introduces an empty test class for testing visualizations. As
implementations start occuring for this class, tests will also be
approrpriately added.

This includes the high level descriptions for each of the methods as
well. Expected use cases will be added to the class description in a
future commit as that gets finalized.

Test Plan: python test/test_quantization.py TestFxModelReportVisualizer

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

vspenubarthi commented:
@pytorchbot merge -g
